lib/src/code_field/code_controller.dart

```diff
@@ -378,6 +379,13 @@ class CodeController extends TextEditingController {
     );
   }
 
+  String _replaceTabsWithSpacesIfNeeded(String text) {
+    if (modifiers.contains(const TabModifier())) {
+      return text.replaceAll(Symbols.tab, Symbols.space * params.tabSpaces);
```

Comment:
I think constants are an overkill here. Tab is clear from `\t`, and a space is clear from multiplying to `tabSpaces`.

Reply:
Constants seem more reliable to me. A space symbol ' ' can be accidentally edited like '  ' and we wouldn't notice the difference.

Reply:
This is why IDEs have monospace fonts. This is why we have tests. And this is why we use `backticks` on GitHub.

Still I guess `\t` is explicit enouth.
And for space do you think we can use `charcode` package if you still want a constant?

`String.fromCharCode($space)`

We already use it elsewhere in our editor.

Reply:
It's just my inexperience, but I didn't know before that '\t' is a tab. `fromCharCode` seems like a complication to me. 

1. I can move the tab and space constants from `Symbols.dart` into the code controller.
2. I can just comply and use '\t' and ' '.

Reply:
Constants should not be created just to move all literals to the top. They are good to:
1. Reuse something that can fall out of sync (like size).
2. Extract something long to make its usage shorter (like a long string).
3. Allow to find all uses of something (like a URL).

Nothing of this applies here.

1. If it is a space, it will always be a space.
2. `' '` is shorter than any constant we can make of it.
3. Searching for "all spaces" is pointless.

Also `const _space = ' '; return _space;` is just as insecure as `return ' ';` but just is more complicated.

So I still prefer `return text.replaceAll('\t', ' ' * params.tabSpaces);`
Also `'\t'` for tab is a common knowledge.

My second best is:
`return text.replaceAll('\t', String.fromCharCode($space) * params.tabSpaces);`
It has the benefit that we know that it is a space because it is defined in a package. No one will fiddle with it replacing a space with an unbreakable space that still looks like `' '`.

My third best is this although I don't like it:
```dart
const _tab = '\t';
const _space = ' ';
```

Reply:
I like the 3rd option after all. It is also easier to read: `replaceAll(_tab, _space * params.tabSpaces)` vs `replaceAll('\t', ' ' * params.tabSpaces)`.